xor_pointers.c

```diff
@@ -0,0 +1,80 @@
+/* xor_lists.c
+ * Jeremy Singer
+ * Example of doubly-linked list data structure
+ * featuring xor'd pointers for shared next/prev field
+ * 
+ * This code inspired by OOPSLA'20 talk:
+ * "Sound Garbage Collection for C using Pointer Provenance"
+ * https://dl.acm.org/doi/10.1145/3428244
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#define NUM_CELLS 100
+#define WORD long int
+
+typedef struct cell
+{
+	WORD data;
+	WORD ptr;
+} cell_t;
+
+cell_t *alloc_cell(WORD payload, cell_t *tail)
+{
+	cell_t *cell = (cell_t *)malloc(sizeof(cell_t));
+	cell->data = payload;
+	/* xor this cell's address into the tail's ptr field */
+	if (tail != NULL)
+		tail->ptr ^= (WORD)cell;
+	/* store tail address in this cell's ptr field
+	 * NOTE most recently allocated cell (i.e. head)
+	 * only stores a single ptr, not an XOR'd pair
+	 */
+	cell->ptr = (WORD)tail;
```

Comment:
If I understand correctly, assuming a non-CHERI system there's no way to tell if this node is the "last" one in a chain, aside from from the implicit state of the system.

On a CHERI system however,  a check on `cell->ptr` being a valid capability pointer reveals  this information "for free".